### PR TITLE
Data structure for the props

### PR DESCRIPTION
**Pull request by ashwinKumar0505**
The reason for change of props data structure is, when the workspace gets too big, the updation of props takes some more time. To reduce the time taken, the props data structure is changed.

### CODE REVIEW

**Comment by sanandnarayan** [comment]
componentID : [props]
componentID: {propID: prop}

**Comment by sanandnarayan** [comment]
before 
`componentID : [props]`
after
`componentID: {propID: prop}`

or
`byID: {propID: {}};
byComponentID: {componentID:[propid1, propid2]}
`


explained here
https://www.loom.com/share/12a24529556542f79c099e150fa61d93

**Comment by ashwinKumar0505** [comment]
I had seen the video bro. I think second approach seems good. I have suggestions, correct me if i am wrong
Can we use prop-name instead of prop-id as keys. Because prop-names will also be unique. There will no duplicate prop-names.
Instead of having byComponentId as another individual property. Can we move it into the component object structure. Each component will have their set of prop-id's. We can fetch the props based on the id's in the component structure.
For Example:

```
Comp-1:{
id:"comp-1",
type:"Button",
parent:"Comp-0",
children:[],
propIds:["prop-1","prop-2"]
}

```
This are just my suggestions

**Comment by sanandnarayan** [comment]
`Props:{
1: {
	byID:{
		prop-KHTASDNFAS : {
			name: “margin”,
			value: “asfasdf,
			derivative’s…
		}
		prop-asfsf {}
	}
	byComponentID:{
		comp-ASFASFDA:[prop-KTSS, ]
	}
}
}
}`

File: src/react-app-env.d.ts [hunk 1]
```diff
@@ -156,10 +156,19 @@ interface IProp {
   derivedFromComponentType: string | null
 }
 
+interface IPropsById {
+  [propId: string]: IProp
+}
+
 interface IPropsByComponentId {
-  [componentId: string]: IProp[]
+  [componentId: string]: string[]
```
Comment (sanandnarayan) [inline src/react-app-env.d.ts:164]: more stricter

File: src/core/models/components/propsOperations.ts [hunk 2]
```diff
@@ -10,43 +10,53 @@ import { ComponentsState } from './components'
...
-  selectedComponentProps.splice(index, 1)
+  const { componentId, propId } = payload
+  const { props } = loadRequired(draftState, componentId)
+
+  const propIdIndex = props.byComponentId[componentId].findIndex(
+    id => id === propId,
+  )
+
+  delete props.byComponentId[propIdIndex]
+  delete props.byId[propId]
```
Comment (sanandnarayan) [inline src/core/models/components/propsOperations.ts]: change to util fn

File: src/core/models/components/propsOperations.ts [hunk 3]
```diff
@@ -94,26 +104,36 @@ export const deleteCustomProp = (
...
+            prop.derivedFromPropName === propName
+          ) {
+            //Remove the prop if its value is empty
+            if (
+              draftState.customComponentsProps.byId[propId].value.length === 0
+            ) {
+              draftState.customComponentsProps.byComponentId[
+                componentId
+              ].splice(index, 1)
+              delete draftState.customComponentsProps.byId[propId]
```
Comment (sanandnarayan) [inline src/core/models/components/propsOperations.ts:150]: ^

File: src/core/models/components/propsExposion.ts [hunk 4]
```diff
@@ -175,24 +142,43 @@ export const unExposeProp = (
...
+          draftState.customComponents = { ...components }
+          draftState.customComponentsProps = { ...props }
+        } else {
+          const { props, components } = deleteCustomPropUtility(
+            component,
+            customPropName,
+            draftState.componentsById[componentsId],
+            draftState.propsById[propsId],
+          )
+          draftState.componentsById[componentsId] = { ...components }
```
Comment (sanandnarayan) [inline src/core/models/components/propsExposion.ts]: change to util fn to add prop in both variables

**Review by sanandnarayan — APPROVED** [review]

**Review by sanandnarayan — APPROVED** [review]

**Pull request merged**